### DISCUSSION TITLE
Improper treatment of booleans in QueryBuilder::toRawSql for Postgres

### DISCUSSION THREAD

**Discussion by bastien-phi** [laravel/framework#48035] [Q&A]
### Laravel Version

10.17

### PHP Version

8.2.8

### Database Driver & Version

Postgresql 14

### Description

When using Postgresql, booleans are transformed into integers in `Builder::toRawSql()`.

I tried to fix it but I'm a bit worried to make a mess to fix that.

`PostgresConnection`does have a valid `escapeBool`method but the proble is higher in the call stack.

When running `User::query()->where('is_admin', true)->toRawSql();`, `toRawSql()` looks like 

```php
    public function toRawSql()
    {
        return $this->grammar->substituteBindingsIntoRawSql(
            $this->toSql(), $this->connection->prepareBindings([true])
        );
    }
```
which resolves in 
```php
    public function toRawSql()
    {
        return $this->grammar->substituteBindingsIntoRawSql(
            $this->toSql(), [1]
        );
    }
```
because of  https://github.com/laravel/framework/blob/19b42097f542a837d3c0e6aad2abaf2b05a6cbee/src/Illuminate/Database/Connection.php#L713-L729
(note that the boolean cast into integer)

@tpetry Any smart idea to fix that ? 

I guess overriding `prepareBindings`and `bindValues` in `PostgresConnection` should do the trick but is not a valid move for Laravel : 
Related explanations on why booleans are casted to integers in Laravel by Taylor : https://github.com/php/php-src/pull/6801#issuecomment-828783810

### Steps To Reproduce

```php
User::query()->where('is_admin', true)->toRawSql();
```
returns `'select * from "users" where "is_admin" = 1'` 
while it should return `'select * from "users" where "is_admin" = true'`

**Comment by tpetry** [comment]
Changing this in Laravel may break some applications. So it is not a good approach to change it.

Casting those values to `int` is not perfect, but it is also not a problem. PostgreSQL is intelligent enough to recognize that you are comparing a `boolean` column to a `0` or `1` value and will automatically cast them. So for almost any application, this is not a problem. I had some code parts in the past where this automatic casting did not work correctly and PG will then fail with an error about incompatible types. If you haven't seen this yet, the `bool -> int` cast is not a problem.

If you've had this problem or insist on having `is_admin = TRUE` in your query, you can use the `whereBool` method from [tpetry/laravel-postgresql-enhanced](https://github.com/tpetry/laravel-postgresql-enhanced)

**Comment by bastien-phi** [comment]
@tpetry unfortunately, postgres does not seed to be able to cast `int -> bool`.

When I try to replay manually an SQL query fetched from `ddRawSql()`, Postgresql complains about the type : 
```
select * from "users" where "is_admin" = 1;
ERROR:  operator does not exist: boolean = integer
LINE 1: select * from "users" where "is_admin" = 1;
                                               ^
HINT:  No operator matches the given name and argument types. You might need to add explicit type casts.
```

Is it a configuration problem in postgres ? I quite don't understant why the int instead of bool will pass in PDO but not in the pgsql console.

**Comment by tpetry** [comment]
I think maybe that was the edge case for me too, not quite sure. As said, you can use my `whereBool` method.

**Comment by bastien-phi** [comment]
Ok. There is some black magic in PDO that I can't understand but boolean columns are sent as boolean to postgres, not integers.

```php
User::query()->where('is_admin', true)->where('id', 1)->get();
// postgres log : 
// execute pdo_stmt_00000003: select * from ""users"" where ""is_admin"" = $1 and ""id"" = $2","parameters: $1 = 't', $2 = '1'
// toRawSql : 
// select * from "users" where "is_admin" = 1 and "id" = 1

User::query()->where('is_admin', 1)->where('id', 1)->get();
// execute pdo_stmt_00000003: select * from ""users"" where ""is_admin"" = $1 and ""id"" = $2","parameters: $1 = 't', $2 = '1'
// select * from "users" where "is_admin" = 1 and "id" = 1

User::query()->where('is_admin', '1')->where('id', '1')->get();
// execute pdo_stmt_00000003: select * from ""users"" where ""is_admin"" = $1 and ""id"" = $2","parameters: $1 = 't', $2 = '1'
// select * from "users" where "is_admin" = '1' and "id" = '1'

User::query()->where('is_admin', 'true')->where('id', '1')->get();
// execute pdo_stmt_00000003: select * from ""users"" where ""is_admin"" = $1 and ""id"" = $2","parameters: $1 = 't', $2 = '1'
// select * from "users" where "is_admin" = 'true' and "id" = '1'
```
It looks like the bindings are treated slightly differently between Eloquent -> PDO -> Database and Eloquent -> rawSql.

In any case, great job on your recent contributions, it will help a lot !

**Comment by CamKem** [comment]
@tpetry - I have just run into this issue today, raw SQL dumps have the bool value as 1/0 - when I take the sql query dump and try to run it in TablePlus because of the native bool types, the query errors out. Will the package you mentioned fix the output for the toRawSql() Method for Postgres?

**Reply by tpetry** [reply]
Yes. the `whereBool()` method of my package fixes it.

But the behaviour in Laravel can't be changed as it would break a lot of apps.
